vllm/spec_decode/multi_step_worker.py

```diff
@@ -11,6 +12,68 @@
 from vllm.worker.worker import Worker
 
 
+class NGramDrafter():
+
+    def __init__(self, device, vocab_size):
+        self.ngram_len = 2
+        self.device = device
+        self.vocab_size = vocab_size
+
+    def spec_infer(
+        self,
+        seq_group_metadata_list: List[SequenceGroupMetadata],
+        num_steps: int,
+    ) -> Optional[List[SamplerOutput]]:
+
+        arr = []
+        has_spec_out = False
+        for seq_group_metadata in seq_group_metadata_list:
+            seq_data = next(iter(seq_group_metadata.seq_data.values()))
+
+            input_ids = torch.from_numpy(
+                np.asarray(seq_data.get_token_ids(), np.int64))
+            ngram_tensor = input_ids[-1 * self.ngram_len:]
+            windows = input_ids.unfold(dimension=0,
+                                       size=self.ngram_len,
+                                       step=1)
+            matches = (windows == ngram_tensor).all(dim=1)
+            match_indices = matches.nonzero(as_tuple=True)[0]
+            if match_indices.size()[0] > 1:
+                has_spec_out = True
+                res = seq_data.get_token_ids()
+                res = res[match_indices[0] + self.ngram_len:match_indices[0] +
+                          self.ngram_len + num_steps]
+                res_len = len(res)
+                # pad 0 towards output as num_steps tokens required
+                for i in range(res_len, num_steps):
+                    res.append(0)
+            else:
+                res = [0] * num_steps
+
+            arr.append(res)
+
+        if not has_spec_out:
+            return None
+
+        outputs = []
+        arr = np.array(arr).transpose()
+        for i in range(num_steps):
+            token_probs = torch.zeros(
+                (len(seq_group_metadata_list), self.vocab_size),
+                dtype=torch.float32).to(self.device)
+
+            token_ids = torch.from_numpy(arr[i]).long().to(self.device)
+            for idx, token in enumerate(arr[i]):
+                token_probs[idx][token] = 1
```

Comment:
I feel that this can be done with tensor ops instead of a for-loop to reduce CPU-GPU data transfer.

Reply:
use scatter_ func to accelerate this func